code/modules/power/proto_sh_emitter.dm

```diff
@@ -0,0 +1,293 @@
+				return FALSE
+
+	return TRUE
+
+/obj/machinery/power/proto_sh_emitter/ui_interact(mob/user, datum/tgui/ui)
+	add_fingerprint(user)
+	if(panel_open)
+		to_chat(user, "<span class='warning'>You need to close the panel first!</span>")
+		return
+	if(!anchored)
+		to_chat(user, "<span class='warning'>You need to anchor the [src] first!</span>")
+		return
+	if(!is_operational)
+		to_chat(user, "<span class='warning'>There is no power in this area!!</span>")
+		return
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "PrototypeShieldEmitter", name)
+		ui.open()
+
+/obj/machinery/power/proto_sh_emitter/ui_data(mob/user)
+	var/list/data = list()
+	data["on"] = is_on
+	data["powered"] = is_operational
+	data["has_barrier"] = has_barrier
+
+	data["width"] = a_o + b_o
+	data["height"] = c_o
+
+	return data
+
+/obj/machinery/power/proto_sh_emitter/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
+
+	switch(action)
+		if("on")
+			is_on = !is_on
+			. = TRUE
+		if("emit")
+			if(is_on && !has_barrier && check_map_borders(a_o, c_o, b_o, c_o))
+				build_barrier(a_i, 2, b_i, c_i, a_o, 1, b_o, c_o, usr)
+				has_barrier = TRUE
+			. = TRUE
+		if("disable")
+			if(has_barrier)
+				remove_barrier(usr)
+				has_barrier = FALSE
+			. = TRUE
+		if("increase_left")
+			a_i++
+			a_o++
+			a_i = clamp(a_i, 0, 1)
+			a_o = clamp(a_o, 1, 2)
+			. = TRUE
+		if("decrease_left")
+			a_i--
+			a_o--
+			a_i = clamp(a_i, 0, 1)
+			a_o = clamp(a_o, 1, 2)
+			. = TRUE
+		if("increase_right")
+			b_i++
+			b_o++
+			b_i = clamp(b_i, 0, 1)
+			b_o = clamp(b_o, 1, 2)
+			. = TRUE
+		if("decrease_right ")
+			b_i--
+			b_o--
+			b_i = clamp(b_i, 0, 1)
+			b_o = clamp(b_o, 1, 2)
+			. = TRUE
+		if("increase_up")
+			c_i++
+			c_o++
+			c_i = clamp(c_i, 2, 4)
+			c_o = clamp(c_o, 3, 5)
+			. = TRUE
+		if("decrease_up")
+			c_i--
+			c_o--
+			c_i = clamp(c_i, 2, 4)
+			c_o = clamp(c_o, 3, 5)
+			. = TRUE
+	update_appearance()
+
+/obj/machinery/holosign/barrier/power_shield
+	name = "powered shield"
+	desc = "A shield to prevent changes of atmospheric and heat transfer"
+	icon = 'icons/effects/effects.dmi'
+	density = FALSE
+	anchored = TRUE
+	CanAtmosPass = ATMOS_PASS_NO
+	resistance_flags = FIRE_PROOF
+	///store the conductivity value of the turf is applyed so that it can be restored on removal
+	var/stored_conductivity = 0
+	///power drain from the apc, in W (so 5000 is 5 kW), per each holosign placed
+	var/obj/item/holosign_creator/shield_projector
+
+/obj/machinery/holosign/barrier/power_shield/Initialize(loc, source_projector)
+	. = ..()
+	if(source_projector)
+		shield_projector = source_projector
+		LAZYADD(shield_projector.signs, src)
+		air_update_turf(TRUE)
+		shield_turf()
+
+/obj/machinery/holosign/barrier/power_shield/Destroy()
+	var/turf/current_turf = loc
+	current_turf.thermal_conductivity = stored_conductivity
+	LAZYREMOVE(shield_projector.signs, src)
+	shield_projector = null
+	return ..()
```

Comment:
What happens if shield_projector is null? or if we never set stored_conductivity to anything?

Reply:
shield projector should never be null since it's the machine that produces those shields
and stored_conductivity is always set to something since every turf has it (if we set to zero it returns zero)